File: packages/adapter-android/app/src/main/kotlin/com/thrw/adapter/android/triggers/SelfCooldown.kt
package com.thrw.adapter.android.triggers

/**
 * ADR 0010 point 1: after thrw initiates a claim or release, it
 * suppresses processing of its own resulting connection-state-changed
 * events for a short window, so it doesn't misread the side effects of
 * its own action as a new trigger.
 *
 * ## Why this became necessary
 *
 * Nothing needed it while the only triggers were `call` and `voip`:
 * telephony state and notification activity are unaffected by thrw's own
 * Bluetooth actions. The **media** triggers (#165/#166) broke that:
 *
 * - thrw disconnects the headset to hand it over, so
 * - audio on that device stops (macOS pauses playback when the output
 *   device disappears), so
 * - the media monitor observes "media ended", and
 * - thrw reports an `event_end` caused entirely by its own action.
 *
 * ## What this deliberately cannot do
 *
 * A time window is a heuristic, not a causal link. Inside the window this
 * cannot distinguish *"audio stopped because we disconnected the
 * headset"* from *"audio stopped because the user pressed pause at that
 * exact moment"* - both are suppressed. The second case means the relay
 * briefly believes a trigger is still active when it isn't; it corrects
 * on the next real state change.
 *
 * That is the trade ADR 0010 chose, and it is worth restating rather than
 * discovering later: **this can suppress genuine user actions that happen
 * to land within ~3s of a claim or release.**
 *
 * ## Scope
 *
 * This is ADR 0010 **point 1 only**. Points 2-5 of that ADR
 * (manual-override detection, scope limitation, known-conflict detection,
 * app-priority allowlist) are separate concerns and deliberately not
 * implemented here.
 *
 * Per ADR 0013 this is **not** a fifth connection state - it is
 * adapter-local bookkeeping layered around `onClaim`/`onRelease`, and
 * nothing in the frozen state machine changes.
 */
class SelfCooldown(
    private val windowMs: Long = DEFAULT_WINDOW_MS,
    private val now: () -> Long = System::currentTimeMillis,
) {
    private var activeUntil = 0L

    /** Called immediately after thrw initiates a claim or release. */
    fun arm() {
        activeUntil = now() + windowMs
    }

    /** Whether trigger reporting should currently be suppressed. */
    fun isActive(): Boolean = now() < activeUntil

    companion object {
        /**
         * 6 seconds, per ADR 0010's 2026-09-23 amendment.
         *
         * `adapter-mac`'s `defaultSelfCooldown` is the same constant on
         * the other side of the same behaviour - change both together.
         *
         * **Was 3 seconds, which was too short (#251).** The old value
         * claimed to be "long enough to cover a Bluetooth profile
         * connect/disconnect and the audio stack settling afterwards".
         * It was not: ADR 0010 *estimated* ~3s, and ADR 0018 later
         * *measured* a claim taking 3-5s to move the route. The window
         * closed before the transition it exists to cover had finished,
         * and that transition's tail was read as a fresh trigger.
         *
         * On hardware, with media playing on both devices, the Mac
         * emitted `media` start/end pairs 1-5 seconds apart while
         * YouTube played continuously. Most-recently-started wins, so
         * each restart took the headset and each end gave it back -
         * bouncing indefinitely, with the relay arbitrating correctly
         * throughout.
         *
         * 6s covers ADR 0018's measured upper bound with margin and
         * stays clear of ADR 0019's 8s command bound. The "short enough
         * that a suppressed genuine action is rare" half of the old
         * reasoning is now carried by exempting `call` as well as
         * `manual_claim` - see [EventKind.bypassesSelfCooldown].
         */
        const val DEFAULT_WINDOW_MS = 6_000L
    }
}

File: packages/adapter-mac/Sources/AdapterMac/Triggers/SelfCooldown.swift
import Foundation

/// 6 seconds, per ADR 0010's 2026-09-23 amendment. `adapter-android`'s
/// `SelfCooldown.DEFAULT_WINDOW_MS` is the same constant on the other
/// side of the same behaviour - change both together.
///
/// **Was 3 seconds, which was too short (#251).** ADR 0010 estimated
/// "~3 seconds" as how long thrw's own side effect takes to play out;
/// ADR 0018 later *measured* a claim taking 3-5s to move the route. So
/// the window closed before the transition it exists to cover had
/// finished, and the tail of that transition was read as a fresh
/// trigger.
///
/// What that looked like on hardware: media playing on both devices,
/// and the Mac emitting `media` start/end pairs 1-5 seconds apart while
/// YouTube played continuously. Since the tie-break is
/// most-recently-started, each restart took the headset back and each
/// end handed it to the phone - bouncing indefinitely, with the relay
/// arbitrating correctly the whole time.
///
/// 6s covers ADR 0018's measured upper bound with margin and stays
/// clear of ADR 0019's 8s command bound, so the two do not interact
/// confusingly.
public let defaultSelfCooldown: Duration = .seconds(6)

/// ADR 0010 point 1: after thrw initiates a claim or release, it
/// suppresses processing of its own resulting connection-state-changed
/// events for a short window, so it doesn't misread the side effects of
/// its own action as a new trigger. Swift mirror of
/// `adapter-android`'s `SelfCooldown.kt` - see that file for the full
/// reasoning.
///
/// Became necessary with the media triggers (#165/#166): thrw
/// disconnects the headset, macOS pauses playback because the output
/// device vanished, and the media monitor reports "media ended" - an
/// event caused entirely by thrw's own action.
///
/// **What it deliberately cannot do**: a time window is a heuristic, not
/// a causal link. Inside the window this cannot tell *"audio stopped
/// because we disconnected"* from *"the user pressed pause at that exact
/// moment"* - both are suppressed. That is the trade ADR 0010 chose.
///
/// Per ADR 0013 this is **not** a fifth connection state - it is
/// adapter-local bookkeeping layered around `onClaim`/`onRelease`.
/// Lock-guarded and `@unchecked Sendable`: ``MacNode`` is `Sendable`
/// (via ``NodeInterface``), and this holds mutable state. The lock makes
/// that genuinely safe rather than merely asserted - `onClaim` arms it
/// from the command task while a trigger monitor reads it from its own.
public final class SelfCooldown: @unchecked Sendable {
    private let lock = NSLock()
    private let window: Duration
    private let now: () -> ContinuousClock.Instant
    private var activeUntil: ContinuousClock.Instant?

    public init(
        window: Duration = defaultSelfCooldown,
        now: @escaping () -> ContinuousClock.Instant = { ContinuousClock.now }
    ) {
        self.window = window
        self.now = now
    }

    /// Called immediately after thrw initiates a claim or release.
    public func arm() {
        lock.lock()
        defer { lock.unlock() }
        activeUntil = now().advanced(by: window)
    }

    /// Whether trigger reporting should currently be suppressed.
    public func isActive() -> Bool {
        lock.lock()
        defer { lock.unlock() }
        guard let activeUntil else { return false }
        return now() < activeUntil
    }
}
